jax/_src/lax/control_flow/loops.py

```diff
@@ -1965,18 +1965,21 @@ def _fori_cond_fun(loop_carry):
 
 @weakref_lru_cache
 def _fori_body_fun(body_fun):
-  body_fun = weakref.ref(body_fun)
+  body_fun_ref = weakref.ref(body_fun)
+
+  @wraps(body_fun)
```

Comment:
Actually, looking more closely here, we probably can't use this mechanism (either `functools.wraps` or `functools.update_wrapper`) because it results in a strong reference of the original function being attached to the wrapped function: https://github.com/python/cpython/blob/aa6579cb60b4fcd652d7bc83fed2668e4ae84db3/Lib/functools.py#L61

Reply:
You are right, then the solution is to inline (parts of) wraps to get the function name to work properly.

Reply:
In the end, I decided to just save the `fun_sourceinfo`, rather than saving all the ingredients (the filename and lineno are read-only).